### ISSUE TITLE
[more.py] Figure desired behavior for queuing and unqueuing

### ISSUE THREAD

**Issue by Androbin** [apertium/phenny#410]
See https://github.com/apertium/phenny/issues/159#issuecomment-373937493 for reference
@jonorthwash /cc

**Comment by Androbin** [comment]
> So it queues infinitely until someone happens to say .more, and then it floods the channel until the [now very large] queue is empty? That doesn't sound like what we want...

@jonorthwash Previously, there were some exceptions to queuing, thus the spamming.
#408 will only print the first message if there are no queued messages.
And at most every 60 seconds, it will print the following message:
"There are <count> queued messages. Type '.more' to view."

Does this sound more like what we want?

**Comment by jonorthwash** [comment]
Does the queue get cleared periodically, or does it constantly fill up?

**Comment by Androbin** [comment]
When a new message is generated:
- If there are queued messages
- - Enqueue everything
- Else:
- - Print the first message
- - Enqueue the rest
- If there was no notification in the last 60 seconds:
- - Print "There are `count` queued messages. Type `.more` to view."
- - Reset timer

When someone types `.more` or `.more <N>` or `.more <N> <tag>`:
- Print the next message(s)
- Print "There are `count` queued messages. Type `.more` to view."
- Reset timer

**Comment by Androbin** [comment]
> Does the queue get cleared periodically, or does it constantly fill up?

It doesn't print more than one message without request to do so.
But it keeps people notified of how many (new) messages there are.

**Comment by Androbin** [comment]
@jonorthwash How about responding to `.more` in a private chat?
This would not spam the original channel in any way...

**Comment by jonorthwash** [comment]
The behaviour you mention is terrible, imho.  (@ftyers should give an opinion too.)

As I see it, we will normally get *no* useful messages in the channel, and instead every 60 seconds get a message from begiak saying there are hundreds of new messages.  This causes more problems then it's worth, and should not be pursued.

**Comment by jonorthwash** [comment]
The only way we'd get useful messages on occasion is for someone to take it upon themselves to clear the `.more` queue.

**Comment by Androbin** [comment]
> instead every 60 seconds get a message from begiak saying there are hundreds of new messages

You did not listen to me. I have always stated explicitly, that the such notification will only be sent, when a new message was generated. The 60 seconds are an additional frequency limit.

**Comment by jonorthwash** [comment]
That's slightly better, but either way we'd still have this issue:
> The only way we'd get useful messages on occasion is for someone to take it upon themselves to clear the `.more` queue.

**Comment by Androbin** [comment]
Let's step through this:
* We can limit the number of messages being generated to begin with
* If we only generate messages that we wanna see, we don't wanna drop any
* When there are multiple messages, we don't wanna print all of them right away
* If we don't print them automatically, someone has to request that manually
* If someone has to clear the queue manually, the channel might be flooded
* We can have status messages (e.g. N new messages) to keep the channel aware
* We don't wanna have to keep track of such information

May I propose the following (possibly as a building block):
Messages are always queued. The head of the queue is un-queued as soon as there has been no message of the same tag for e.g. 5 minutes (which can be directly afterwards).

**Comment by Androbin** [comment]
Now that I think about it, this should be the throttling option suggested by @sushain97.

**Comment by sushain97** [comment]
> The head of the queue is un-queued as soon as there has been no message (of the same tag) for e.g. 5 minutes (which can be directly afterwards).

This is actually debouncing.

**Comment by Androbin** [comment]
> Debouncing will bunch a series of sequential calls to a function into a single call to that function. It ensures that one notification is made for an event that fires multiple times.

Doesn't that mean that debouncing would post multiple messages as one message?

**Comment by sushain97** [comment]
> Doesn't that mean that debouncing would post multiple messages as one message?

No. Traditionally, debouncing would post *only* the last message (trailing edge). However, we can include a summarizing function that would compress the messages instead.

**Comment by Androbin** [comment]
Now we're talking. How could that look like, given that IRC doesn't allow line breaks?

Actually, wouldn't it be fun to do newlines by padding the message with whitespace, based on the line width of the web interface? :sweat_smile:
